### PR TITLE
finagle-example: adds an HTTP java example.

### PR DESCRIPTION
**Pull request by leonmaia**
Problem

Lack of examples to create a Http server using java.

Solution

Build a very minimal example showing how to create a server, some filters, and a very basic routing.

Result

More examples of how to start a java project using Finagle.

### CODE REVIEW

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/CatService.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.twitter.finagle.example.java.http;
+
+public final class CatService {
+    static CatsDB db = new CatsDB();
```
Comment (kevinoliver) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/CatService.java]: can we mark this `private final`? good to have good code hygiene even in sample code.

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/CatsDB.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package com.twitter.finagle.example.java.http;
+
+import java.util.HashMap;
+
+public class CatsDB {
+    static HashMap<Integer, Cat> db = new HashMap<>();
```
Comment (kevinoliver) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/CatsDB.java]: same `private final` here too.

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/HandleErrors.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import com.twitter.finagle.SimpleFilter;
+import com.twitter.finagle.http.Request;
+import com.twitter.finagle.http.Response;
+import com.twitter.util.ExceptionalFunction;
+import com.twitter.util.Future;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+
+/**
+ * A simple Finagle that intercepts Exceptions and converts them to a more comprehensible HTTP Response.
```
Comment (kevinoliver) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/HandleErrors.java]: s/Finagle/Filter/

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/HandleErrors.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            @Override
+            public Response applyE(Throwable in) {
+                Response resp = Response.apply();
+                if (in instanceof NumberFormatException) {
+                    resp.setStatus(HttpResponseStatus.BAD_REQUEST);
+                    resp.setContent(ChannelBuffers.wrappedBuffer(in.getMessage().getBytes()));
+
+                    return resp;
+                }
+                in.printStackTrace();
```
Comment (kevinoliver) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/HandleErrors.java]: feel free to ignore, but i prefer not teaching people to log, so can we omit this line?

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/LoggingFilter.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import com.twitter.util.Stopwatch$;
+
+/**
+ * A simple Finagle filter that logs the Request total time in milliseconds.
+ */
+public final class LoggingFilter extends SimpleFilter<Request, Response> {
+
+    @Override
+    public Future<Response> apply(Request req, Service<Request, Response> service) {
+        final Duration e = Stopwatch$.MODULE$.start().apply();
```
Comment (kevinoliver) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/LoggingFilter.java]: bummer that we don't have a nicer java API for `Stopwatch`. i'll try to address that for the future.

that said, you should capture the `Elapsed` from calling `start()`. then call `apply()` inside the `map` function below.

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/HttpServer.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import com.twitter.finagle.http.Request;
+import com.twitter.finagle.http.Response;
+import com.twitter.util.Await;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+
+public class HttpServer {
+
+    static InetSocketAddress addr = new InetSocketAddress(InetAddress.getLoopbackAddress(), 8888);
```
Comment (kevinoliver) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/HttpServer.java]: `private final` here too please.

**Comment by kevinoliver** [comment]
@leonmaia this is great — thanks! i just have a few small fixes i'd like to see applied and then would be happy to pull this in.

**Comment by mosesn** [comment]
Hey @leonmaia, do you think you'll have a chance to get to this sometime this week?  No rush if you're busy 😀

**Comment by leonmaia** [comment]
Thanks for the comments @kevinoliver :)

@mosesn sure! Sorry for taking so long :) I was very busy last week.

**Comment by mosesn** [comment]
No worries, just wanted to make sure it didn't fall off your plate.

**Comment by leonmaia** [comment]
I've applied all changes @kevinoliver, please let me know if there's something else. :)

**Comment by kevinoliver** [comment]
@leonmaia lgtm i'll work on getting this merged in now. thanks!

File: finagle-example/src/main/java/com/twitter/finagle/example/java/http/HandleErrors.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            public Response applyE(Throwable in) {
+                Response resp = Response.apply();
+                if (in instanceof NumberFormatException) {
+                    resp.setStatus(HttpResponseStatus.BAD_REQUEST);
+                    resp.setContent(ChannelBuffers.wrappedBuffer(in.getMessage().getBytes()));
+
+                    return resp;
+                }
+                resp.setStatus(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+                resp.setContent(ChannelBuffers.wrappedBuffer(in.getMessage().getBytes()));
```
Comment (luciferous) [inline finagle-example/src/main/java/com/twitter/finagle/example/java/http/HandleErrors.java]: I'm surprised this works – I thought finagle-http API was Netty free. Is this about Java subverting Scala access modifiers?

**Comment by mosesn** [comment]
@luciferous yes.  HttpMessageProxy extends Proxy, which makes the netty Message methods available.

**Comment by kevinoliver** [comment]
Just merged this locally. Should show up on the `develop` branch this Monday. Thanks again.

**Pull request closed without merging**